### PR TITLE
Add nist_cvss_validation flag to Flaw API and BBSync, update BZImport

### PR DESCRIPTION
**Pull request by jsvob**
Follow-up to https://github.com/RedHatProductSecurity/osidb/pull/279.

This PR updates the Flaw API to propagate the recently added `nist_cvss_validation` field in the Flaw model, adds `nist_cvss_validation` flag sync to BBSync, and switches BZImport to import the `nist_cvss_validation` flag to the new Flaw field `nist_cvss_validation` instead of `FlawMeta`.

Closes OSIDB-1006

### CODE REVIEW

File: apps/bbsync/query.py [hunk 1]
```diff
@@ -220,11 +220,12 @@ def generate_keywords(self):
     def generate_flags(self):
         """
         generate query for Bugzilla flags
-        TODO: needinfo, nist_cvss_validation and other flags
+        TODO: needinfo and other flags
         """
         self._query["flags"] = []
         self.generate_hightouch_flags()
         self.generate_requires_doc_text_flag()
+        self.generate_nist_cvss_validation_flag()
```
Comment (jobselko) [inline apps/bbsync/query.py:228]: Nitpick: the order of these methods (lines 226-228 in this file and lines 831-833 in `convertors.py)` does not match the order of their implementation (lines 230-281 in this file). I would prefer to have it consistent for easier readability.

File: apps/bbsync/query.py [hunk 2]
```diff
@@ -237,10 +238,8 @@ def generate_hightouch_flags(self):
             Flaw.FlawMajorIncident.CISA_APPROVED: ("-", "+"),
             # flags NOVALUE and INVALID are ignored
         }
-
         if self.flaw.major_incident_state in flags_to_write:
             hightouch, hightouch_lite = flags_to_write[self.flaw.major_incident_state]
-
```
Comment (jobselko) [inline apps/bbsync/query.py:243]: You probably deleted these lines by mistake.

File: apps/bbsync/query.py [hunk 3]
```diff
@@ -262,6 +261,24 @@ def generate_requires_doc_text_flag(self):
...
+
+        flag_to_write = {
+            Flaw.FlawNistCvssValidation.REQUESTED: "?",
+            Flaw.FlawNistCvssValidation.APPROVED: "+",
+            Flaw.FlawNistCvssValidation.REJECTED: "-",
+            # flag NOVALUE is ignored
+        }
+
+        if self.flaw.nist_cvss_validation in flag_to_write:
+            flag_value = flag_to_write[self.flaw.nist_cvss_validation]
```
Comment (jobselko) [inline apps/bbsync/query.py:277]: You can use the ":=" operator, see `generate_requires_doc_text_flag()`.

File: apps/bbsync/tests/test_query.py [hunk 4]
```diff
@@ -1182,3 +1186,37 @@ def test_generate_requires_doc_text(
...
+            # flag to ignore
+            (Flaw.FlawNistCvssValidation.NOVALUE, None, False),
+        ],
+    )
+    def test_generate_nist_cvss_validation(
+        self, field_state, flag_value, should_convert
+    ):
+        """
+        Tests that nist_cvss_validation field is correctly converted into a flag.
+        Other flag-producing fields are set not to produce flags for ease of testing.
```
Comment (jobselko) [inline apps/bbsync/tests/test_query.py:1206]: Nitpick: If you want, you can use the code below to avoid this note:

        if should_convert:
            assert {"name": "nist_cvss_validation", "status": flag_value} in flags
        else:
            assert {"name": "nist_cvss_validation", "status": flag_value} not in flags

File: collectors/bzimport/convertors.py [hunk 5]
```diff
@@ -623,6 +623,28 @@ def requires_summary(self):
...
+        mapping = {
+            "": Flaw.FlawNistCvssValidation.NOVALUE,
+            "?": Flaw.FlawNistCvssValidation.REQUESTED,
+            "+": Flaw.FlawNistCvssValidation.APPROVED,
+            "-": Flaw.FlawNistCvssValidation.REJECTED,
+        }
+
+        # If the flag is a different character, let this crash. Shouldn't
+        # happen though, so no special handling. BZ docs specify only ?+-
+        # https://www.bugzilla.org/docs/2.18/html/flags-overview.html
```
Comment (jobselko) [inline collectors/bzimport/convertors.py:645]: I am unsure if this note is necessary. As you mentioned, there should be only values `""`/`"?"`/`"+"`/`"-"`, so no need to differentiate between `[]` and `get().`

File: osidb/tests/test_endpoints.py [hunk 6]
```diff
@@ -1095,6 +1097,7 @@ def test_get_flaw_with_token(
         flaw1 = FlawFactory.build(
             major_incident_state=Flaw.FlawMajorIncident.APPROVED,
             requires_summary=Flaw.FlawRequiresSummary.APPROVED,
+            nist_cvss_validation=Flaw.FlawNistCvssValidation.REQUESTED,
```
Comment (jobselko) [inline osidb/tests/test_endpoints.py:1100]: Nitpick: No need to add `nist_cvss_validation` into this test, as the original test did not mention it.

You test `nist_cvss_validation` in `test_get_flaw()` and that shoud be sufficient.

File: apps/bbsync/tests/test_query.py [hunk 7]
```diff
@@ -1134,10 +1134,12 @@ def test_generate_hightouch_and_hightouch_lite(
         """
         Tests that major_incident_state is correctly converted into hightouch and
         hightouch-lite flags.
+        Other flag-producing fields are set not to produce flags for ease of testing.
```
Comment (jobselko) [inline apps/bbsync/tests/test_query.py:1137]: Changes of this type should be in a different commit. Also please see a comment in `test_generate_nist_cvss_validation()`, it might be useful.

File: collectors/bzimport/tests/test_convertors.py [hunk 8]
```diff
@@ -233,6 +234,7 @@ def test_basic(self):
         assert flaw.package_versions.first() == package_versions
         assert flaw.major_incident_state == Flaw.FlawMajorIncident.REQUESTED
         assert flaw.requires_summary == Flaw.FlawRequiresSummary.NOVALUE
+        assert flaw.nist_cvss_validation == Flaw.FlawNistCvssValidation.REJECTED
```
Comment (jobselko) [inline collectors/bzimport/tests/test_convertors.py:237]: I think there is an issue in `FlawFactory`. The `nist_cvss_validation` field is not conditional on the `cvss` field. This way, you can create a flaw without `cvss` but with `nist_cvss_validation=REJECTED`, which is not correct. 

Btw. `requires_summary` probably has a similar problem.

Comment (jobselko) [reply collectors/bzimport/tests/test_convertors.py:237]: This should be solved in a different PR. You can ignore the comment above.

File: apps/bbsync/query.py (hunk 1 above)
Comment (jsvob) [reply apps/bbsync/query.py:228]: I don't understand.

query.py:
```
226:        self.generate_hightouch_flags()
227:        self.generate_requires_doc_text_flag()
228:        self.generate_nist_cvss_validation_flag()
230:    def generate_hightouch_flags(self):
248:    def generate_requires_doc_text_flag(self):
264:    def generate_nist_cvss_validation_flag(self):
```
- generate_hightouch_flags() reads from flaw.major_incident_state
- generate_requires_doc_text_flag() reads from flaw.requires_summary
- generate_nist_cvss_validation_flag() reads from flaw.nist_cvss_validation

convertors.py:
```
831:            major_incident_state=self.major_incident_state,
832:            requires_summary=self.requires_summary,
833:            nist_cvss_validation=self.nist_cvss_validation,
```

Therefore all triplets of lines/blocks are in the same order:
- query.py 226 & query.py 230-246 & convertors.py 831
- query.py 227 & query.py 248-262 & convertors.py 832
- query.py 228 & query.py 264-280 & convertors.py 833

Please provide an actionable explanation of the code change you envision.

File: apps/bbsync/query.py (hunk 2 above)
Comment (jsvob) [reply apps/bbsync/query.py:243]: Yes. I thought it was `black` but it was me.

File: collectors/bzimport/convertors.py (hunk 5 above)
Comment (jsvob) [reply collectors/bzimport/convertors.py:645]: I'll change it to match `requires_summary()`.

File: apps/bbsync/tests/test_query.py (hunk 7 above)
Comment (jsvob) [reply apps/bbsync/tests/test_query.py:1137]: I'll move the change to a separate commit.

**Comment by jsvob** [comment]
@jobselko Thank you for the review. I replied to your feedback and fixed code where applicable. There are 2 commits instead of 1 commit now. Additionally, added a migration which I forgot to push.

File: osidb/tests/test_endpoints.py (hunk 6 above)
Comment (jobselko) [reply osidb/tests/test_endpoints.py:1100]: > Nitpick: No need to add `nist_cvss_validation` into this test, as the original test did not mention it.

- This sentence is under 1100 line and it belongs to the `test_get_flaw_with_token()` test.
- In the past, I added `major_incident_state` and `requires_summary` into `test_get_flaw()` and `test_get_flaw_with_token()` because both these tests had contained `is_major_incident` (this was replaced by `major_incident_state)` and `FlawMetaFactory` of the `REQUIRES_SUMMARY` type (this was replaced by  `requires_summary`).

> You test `nist_cvss_validation` in `test_get_flaw()` and that shoud be sufficient.

- None of the tests in `test_endpoints` contained `FlawMeta`'s `NIST_CVSS_VALIDATION,` but I think putting it into one test (e.g. into `test_get_flaw()`) is ok.

Does it make sense to you now? :)

File: apps/bbsync/query.py (hunk 1 above)
Comment (jobselko) [reply apps/bbsync/query.py:228]: Oh, I am sorry! I am looking at it again, and you are right. I made a mistake, sorry for the noise.

**Comment by jobselko** [comment]
@jsvob Could you please split the `Add nist_cvss_validation flag to Flaw API and BBSync, update BZImport` commit into three parts (BBSync, BZImport, and API)?
Also, the new migration file you added should be in a separate commit (I have not reviewed it yet.)

File: osidb/tests/test_endpoints.py (hunk 6 above)
Comment (jsvob) [reply osidb/tests/test_endpoints.py]: Fixed & committed.

**Review by osoukup — COMMENTED** [review]
LGTM. Great comprehensive job with the migration and everything!

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+memory issues and runtime efficiency, the migration uses a generator,
+islice and bulk_create as recommended by
+https://docs.djangoproject.com/en/3.2/ref/models/querysets/#bulk-create
+
+This migration should happen in lockstep with adding bzimport for Flaw
+field nist_cvss_validation. This ensures that any future operations performed
+using the new field do not result in data loss. (For a hypothetical example,
+if the flag value was recorded only in FlawMeta but not in the Flaw field,
+bbsync based on the field would effectively delete the flag from bugzilla,
+and that would also delete the FlawMeta-based flag on next bzimport.)
```
Comment (jobselko) [inline osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py:17]: As we already discussed, this situation currently cannot happen, so this paragraph evokes an unrealistic issue (which imho is not good).

I am not against migrations, but having them included in the commit `Add nist_cvss_validation to Flaw BZImport.` seems inappropriate. The commit name says nothing about migrations. If we stick to the fact that we should make them together with BZImport at the latest, it also implies that we can make them before the BZImport. So a nitpick from my side is to have the following commits: 1. migration, 2. bzimport, 3. bbsync, 4. api.

To be precise, migrations should be done together with a new model (if we want to do them now and not at the end of Basic Workflow).

**Review by jobselko — CHANGES_REQUESTED** [review]
Just a suggestion, but maybe it would be easier to create a new PR for migrations.

@jsvob Have you tried to run the migrations locally? I am struggling a bit with it.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        bulk_update(apps, batch_flaws_flags)
+
+
+def forwards_func(apps, schema_editor):
+    """
+    For all Flaws that have the bugzilla flag nist_cvss_validation set, this
+    migration sets the value of the field nist_cvss_validation to be identical
+    to existing FlawMeta NIST_CVSS_VALIDATION.
+    """
+    batch_size = 1000
```
[the remark this answers is not in the document]
Comment (osoukup) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py:67]: I assume it is mean that the name should be uppercase.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+    run_batches(apps, batch_size, generate_flag_values(apps))
+
+
+def backwards_func(apps, schema_editor):
```
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py:72]: Because it's documented in https://docs.djangoproject.com/en/3.2/ref/migration-operations/. If we never want reversibility, please discuss with project owner, but seeing as he's already approved, I'm not going to engage in this proactively.

For `FlawAcknowledgment` see https://github.com/RedHatProductSecurity/osidb/blob/master/osidb/migrations/0082_migrate_flawacknowledgment_flawmeta.py#L69-L76.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py:67]: Yes, I'll fix it so that it's a constant. Thanks.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def run_batches(apps, batch_size, flag_val_generator):
+    """
+    Slices all Flaws that should be updated into batches and executes
+    the bulk Flaw model updates.
+    """
+    while True:
+        # Limiting memory usage by using a generator and slicing recommended in
+        # https://docs.djangoproject.com/en/3.2/ref/models/querysets/#bulk-create
```
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py:54]: I'll remove the redundant ones then. Thanks.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def bulk_update(apps, flaws_flags):
+    """
+    Updates flag nist_cvss_validation on Flaws provided in the argument.
+    """
+    Flaw = apps.get_model("osidb", "Flaw")
+    flaws = []
+    for flaw, flag in flaws_flags:
+        flaw.nist_cvss_validation = flag
```
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py:42]: As for your 1st paragraph: You are right. I'll fix it. Thank you for noticing this!

As for your 2nd paragraph: See discussion thread https://github.com/RedHatProductSecurity/osidb/pull/282#discussion_r1276171578

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py (hunk 9 above)
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py:17]: Also, `this situation currently cannot happen` is not applicable to this PR and this migration.

The `situation currently cannot happen` that we did discuss was about submitting PRs that enable writable API in the timeframe before enabling write access on the production instance.

This migration addresses a different problem that would happen on the read-only production instance. The problem and resolution is described using a hypothetical example because that's IMO the shortest explanation possible.

**Comment by jsvob** [comment]
Using the buggy and improperly tested migration, the resulting Flaw doesn't validate:

```
>>> FlawMeta.objects.filter(type="NIST_CVSS_VALIDATION").first().flaw.clean_fields()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 1276, in clean_fields
    raise ValidationError(errors)
django.core.exceptions.ValidationError: {'nist_cvss_validation': ["Value '+' is not a valid choice."]}
>>> 
```

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py]: Fixed.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py]: Fixed.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py]: Fixed.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py (hunk 9 above)
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py:17]: After a 1:1, I'll delete the comment.

**Review by jobselko — APPROVED** [review]
Discussed on 1:1.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py]: Discussed. No action.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+def forwards_func(apps, schema_editor):
+    """
+    For all Flaws that have the bugzilla flag nist_cvss_validation set, this
+    migration sets the value of the field nist_cvss_validation to be identical
+    to existing FlawMeta NIST_CVSS_VALIDATION.
+    """
+    batch_size = 1000
+
+    run_batches(apps, batch_size, generate_flag_values(apps))
```
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py]: Modified based on our 1:1.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
+"""
+Written manually on 2023-07-17
+
+Copy FlawMeta NIST_CVSS_VALIDATION into Flaw field nist_cvss_validation.
+
+To prevent out of memory issues, FlawMeta are iterated using .iterator()
+so that FlawMeta instances are not cached. To prevent further out of
+memory issues and runtime efficiency, the migration uses a generator,
+islice and bulk_create as recommended by
+https://docs.djangoproject.com/en/3.2/ref/models/querysets/#bulk-create
```
[the remark this answers is not in the document]
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py]: Modified based on our 1:1.

File: osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py (hunk 9 above)
Comment (jsvob) [reply osidb/migrations/0086_migrate_flaw_flag_nist_cvss_validation.py]: Modified based on our 1:1.
